storage-node/packages/colossus/paths/asset/v0/{id}.js

```diff
@@ -54,24 +69,36 @@ module.exports = function (storage, runtime, ipfsHttpGatewayUrl, anonymous) {
 
       const id = req.params.id // content id
 
-      // First check if we're the liaison for the name, otherwise we can bail
-      // out already.
+      // Check if we're the liaison for the content
       const roleAddress = runtime.identities.key.address
       const providerId = runtime.storageProviderId
       let dataObject
       try {
-        debug('calling checkLiaisonForDataObject')
         dataObject = await runtime.assets.checkLiaisonForDataObject(providerId, id)
-        debug('called checkLiaisonForDataObject')
       } catch (err) {
         errorHandler(res, err, 403)
         return
       }
 
-      const sufficientBalance = await runtime.providerHasMinimumBalance(3)
+      // Early filtering on content_length..do not wait for fileInfo
+      // ensure its less than max allowed by node policy.
+      const filterResult = filter({}, req.headers)
+
+      if (filterResult.code !== 200) {
+        errorHandler(res, new Error(filterResult.message), filterResult.code)
+        return
+      }
+
+      // Ensure content_length from request equals size in data object.
+      if (!dataObject.size_in_bytes.eq(filterResult.content_length)) {
+        errorHandler(res, new Error('Content Length does not match expected size of content'), 403)
+        return
+      }
+
+      const sufficientBalance = await runtime.providerHasMinimumBalance(PROCESS_UPLOAD_BALANCE)
```

Comment:
`providerHasMinimumBalance` checks for the free balance rather than the usable balance. Free balance doesn't consider locks. Do we want such behavior?

Reply:
Yes indeed, because node will send out transactions to perform its job. Will update to look at "usable" balance.

Reply:
Looks like our own internal method name is badly chosen but it does in-fact get the spendable/usable balance which is derived from freeBalance - lockedBalance: [ref](https://github.com/polkadot-js/api/blob/23f8eb9713c7edc1118bbab59043238fc84dcb74/packages/api-derive/src/balances/all.ts#L57)

I will change our method name to properly reflect this.

Reply:
done in 42eb0d191e4d43b03f44f93218eec826c1811788